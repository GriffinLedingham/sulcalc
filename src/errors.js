export class NoPokemonError extends RangeError {
  constructor() {
    super("Pokemon cannot be (No Pokemon)");
    this.name = "NoPokemonError";
  }
}

export class NoMoveError extends RangeError {
  constructor() {
    super("Move cannot be (No Move)");
    this.name = "NoMoveError";
  }
}
