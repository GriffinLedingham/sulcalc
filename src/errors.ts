export class NoPokemonError extends RangeError {
  public constructor() {
    super("Pokemon cannot be (No Pokemon)");
    this.name = "NoPokemonError";
  }
}

export class NoMoveError extends RangeError {
  public constructor() {
    super("Move cannot be (No Move)");
    this.name = "NoMoveError";
  }
}
