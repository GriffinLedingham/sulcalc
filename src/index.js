export sulcalc from "./sulcalc";
export Multiset from "./Multiset";
export Pokemon from "./Pokemon";
export Ability from "./Ability";
export Item from "./Item";
export Move from "./Move";
export Field from "./Field";
export * as info from "./info";
export * from "./errors";
export * from "./utilities";
