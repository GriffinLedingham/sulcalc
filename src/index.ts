import * as info from "./info";

export { info };
export { default as sulcalc } from "./sulcalc";
export { default as Multiset } from "./Multiset";
export { default as Pokemon } from "./Pokemon";
export { default as Ability } from "./Ability";
export { default as Item } from "./Item";
export { default as Move } from "./Move";
export { default as Field } from "./Field";
export * from "./errors";
export * from "./utilities";
