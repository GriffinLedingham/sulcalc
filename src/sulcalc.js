import Pokemon from "./Pokemon";
import Move from "./Move";
import Field from "./Field";
import Multiset from "./Multiset";
import { Gens, Stats, Statuses, Types, Weathers } from "./utilities";
import { isPhysicalType, natureMultiplier, effectiveness } from "./info";
import { NoPokemonError, NoMoveError } from "./errors";
import calculate from "./calculate";
import endOfTurn from "./endOfTurn";

const weatherMessages = {
  [Weathers.HAIL]: "in Hail",
  [Weathers.RAIN]: "in Rain",
  [Weathers.SAND]: "in Sand",
  [Weathers.SUN]: "in Sun",
  [Weathers.HEAVY_RAIN]: "in Heavy Rain",
  [Weathers.HARSH_SUN]: "in Harsh Sun",
  [Weathers.STRONG_WINDS]: "in Strong Winds"
};

const statusMessages = {
  [Statuses.POISONED]: "poisoned",
  [Statuses.BADLY_POISONED]: "badly poisoned",
  [Statuses.BURNED]: "burned",
  [Statuses.PARALYZED]: "paralyzed",
  [Statuses.ASLEEP]: "asleep",
  [Statuses.FROZEN]: "frozen"
};

export default function sulcalc(attacker, defender, move, field) {
  attacker = new Pokemon(attacker);
  defender = new Pokemon(defender);
  move = new Move({ ...move, user: attacker, target: defender });
  field = new Field(field);

  const reportPokes = [];
  const reportDamage = [];
  const reportResult = [];

  if (attacker.name === "(No Pokemon)" || defender.name === "(No Pokemon)") {
    throw new NoPokemonError();
  }

  if (move.name === "(No Move)") {
    throw new NoMoveError();
  }

  if (
    field.magicRoom ||
    attacker.ability.name === "Klutz" ||
    (defender.ability.name === "Unnerve" && attacker.item.isBerry())
  ) {
    attacker.item.disabled = true;
  }
  if (
    field.magicRoom ||
    defender.ability.name === "Klutz" ||
    (attacker.ability.name === "Unnerve" && defender.item.isBerry())
  ) {
    defender.item.disabled = true;
  }

  if (
    attacker.ability.name === "Air Lock" ||
    defender.ability.name === "Air Lock"
  ) {
    field.airLock = true;
  }

  if (
    attacker.ability.name === "Fairy Aura" ||
    defender.ability.name === "Fairy Aura"
  ) {
    field.fairyAura = true;
  }

  if (
    attacker.ability.name === "Dark Aura" ||
    defender.ability.name === "Dark Aura"
  ) {
    field.darkAura = true;
  }

  if (
    attacker.ability.name === "Aura Break" ||
    defender.ability.name === "Aura Break"
  ) {
    field.auraBreak = true;
  }

  const effects = endOfTurn(attacker, defender, field);
  effects.messages.unshift("");
  effects.values.unshift(0);

  if (
    (attacker.ability.ignoresAbilities() || move.ignoresAbilities()) &&
    defender.ability.isIgnorable()
  ) {
    defender.ability.disabled = false;
  }

  if (defender.hasCritArmor() || !move.canCrit()) {
    move.critical = false;
  } else if (
    attacker.ability.name === "Merciless" &&
    (defender.isPoisoned() || defender.isBadlyPoisoned())
  ) {
    move.critical = true;
  }

  const dmg = calculate(attacker, defender, move, field);

  if (move.name === "Smack Down" || move.name === "Thousand Arrows") {
    defender.grounded = true;
  }

  attacker.item.disabled = false;
  attacker.item.used = false;
  defender.item.disabled = false;
  defender.item.used = false;
  defender.ability.disabled = false;

  const maxHp = defender.stat(Stats.HP);

  const minPercent = Math.round((dmg[0].min() / maxHp) * 1000) / 10;
  const maxPercent = Math.round((dmg[0].max() / maxHp) * 1000) / 10;

  const convertToDamage = v => Math.max(0, maxHp - v);
  let initDmg = defender.currentHpRange.map(convertToDamage);
  let initDmgBerry = defender.currentHpRangeBerry.map(convertToDamage);

  let moveType, movePower;
  if (move.name === "Hidden Power") {
    moveType = Move.hiddenPowerType(attacker.ivs, field.gen);
    movePower = Move.hiddenPowerBp(attacker.ivs, field.gen);
  } else if (move.name === "Weather Ball") {
    moveType = Move.weatherBall(field.effectiveWeather());
    // gen 3 multiplies damage, not BP
    movePower = field.gen >= Gens.HGSS && moveType ? 100 : 50;
  } else if (move.zMove) {
    moveType = move.type();
    movePower = move.power();
  } else {
    moveType = move.type();
    movePower = null;
  }

  let a, d;
  if (move.isPsyshockLike()) {
    a = Stats.SATK;
    d = Stats.DEF;
  } else if (
    (field.gen >= Gens.HGSS && move.isPhysical()) ||
    (field.gen < Gens.HGSS && isPhysicalType(moveType))
  ) {
    a = Stats.ATK;
    d = Stats.DEF;
  } else {
    a = Stats.SATK;
    d = Stats.SDEF;
  }

  if (attacker.boosts[a]) {
    reportPokes.push((attacker.boosts[a] > 0 ? "+" : "") + attacker.boosts[a]);
  }

  if (field.gen >= Gens.ADV || attacker.evs[a] < 252) {
    const mult =
      field.gen >= Gens.ADV ? natureMultiplier(attacker.nature, a) : 0;
    if (mult) {
      reportPokes.push(attacker.evs[a] + (mult > 0 ? "+" : "-"));
    } else {
      reportPokes.push(attacker.evs[a]);
    }
    reportPokes.push(a === Stats.ATK ? "Atk" : "SpA");
  }

  if (field.gen >= Gens.GSC && !attacker.isItemRequired()) {
    reportPokes.push(attacker.item.name);
  }

  if (field.gen >= Gens.ADV && attacker.ability.name !== "(No Ability)") {
    reportPokes.push(attacker.ability.name);
  }

  reportPokes.push(attacker.name);

  if (!attacker.isHealthy()) {
    reportPokes.push(`(${statusMessages[attacker.status]})`);
  }

  if (move.critical) {
    reportPokes.push("critical hit");
  }

  if (attacker.helpingHand) {
    reportPokes.push("Helping Hand");
  }

  reportPokes.push(move.name);

  if (move.zMove || (move.name === "Hidden Power" && field.gen <= Gens.B2W2)) {
    reportPokes.push(`[${movePower} BP]`);
  } else if (move.hitsMultipleTimes() && move.numberOfHits >= 1) {
    const n = move.numberOfHits;
    reportPokes.push(n > 1 ? `[${n} hits]` : `[${n} hit]`);
  }

  reportPokes.push("vs.");

  if (defender.boosts[d]) {
    reportPokes.push((defender.boosts[d] > 0 ? "+" : "") + defender.boosts[d]);
  }

  if (
    field.gen >= Gens.ADV ||
    defender.evs[d] < 252 ||
    defender.evs[Stats.HP] < 252
  ) {
    reportPokes.push(defender.evs[Stats.HP], "HP /");
    const mult =
      field.gen >= Gens.ADV ? natureMultiplier(defender.nature, d) : 0;
    if (mult) {
      reportPokes.push(defender.evs[d] + (mult > 0 ? "+" : "-"));
    } else {
      reportPokes.push(defender.evs[d]);
    }
    reportPokes.push(d === Stats.DEF ? "Def" : "SpD");
  }

  if (field.gen >= Gens.GSC && !defender.isItemRequired()) {
    reportPokes.push(defender.item.name);
  }

  if (field.gen >= Gens.ADV && defender.ability.name !== "(No Ability)") {
    reportPokes.push(defender.ability.name);
  }

  reportPokes.push(defender.name);

  if (!defender.isHealthy()) {
    reportPokes.push(`(${statusMessages[defender.status]})`);
  }

  const defenderHpPercent = Math.floor(
    (100 * defender.currentHp) / defender.stat(Stats.HP)
  );
  if (defenderHpPercent < 100) {
    reportPokes.push(`at ${defenderHpPercent}%`);
  }

  if (defender.reflect && d === Stats.DEF) {
    reportPokes.push("behind Reflect");
  } else if (defender.lightScreen && d === Stats.SDEF) {
    reportPokes.push("behind Light Screen");
  }

  if (weatherMessages[field.weather]) {
    reportDamage.push(" " + weatherMessages[field.weather]);
  }

  if (dmg[0].min() === dmg[0].max()) {
    reportDamage.push(String(dmg[0].min()), `(${minPercent}%)`);
  } else {
    reportDamage.push(
      `${dmg[0].min()} - ${dmg[0].max()}`,
      `(${minPercent} - ${maxPercent}%)`
    );
  }

  // remove field hazards from current HP for probability calculation
  if (defender.ability.name !== "Magic Guard") {
    let hazardsDmg = 0;
    if (defender.stealthRock) {
      const [numerator, denominator] = effectiveness(
        Types.ROCK,
        defender.types(),
        {
          gen: field.gen
        }
      );
      hazardsDmg += Math.trunc((maxHp * numerator) / (denominator * 8));
    }
    if (defender.spikes > 0 && defender.isGrounded(field)) {
      hazardsDmg += Math.trunc(maxHp / (10 - 2 * defender.spikes));
    }
    initDmg = initDmg.map(v => Math.min(maxHp, v + hazardsDmg));
    initDmgBerry = initDmgBerry.map(v => Math.min(maxHp, v + hazardsDmg));
  }

  const chances = chanceToKo(defender, dmg, {
    initDmgRange: initDmg,
    initDmgRangeBerry: initDmgBerry,
    effects: effects.values,
    rechargeTurns: Number(move.requiresRecharge()),
    toxicCounter: defender.toxicCounter
  });

  effects.messages = effects.messages.slice(1);
  effects.values = effects.values.slice(1);

  const displayChances = [];
  for (let i = 0; i < chances.roundedChances.length; i++) {
    const nhko = `${i ? i + 1 : "O"}HKO`;
    const fractionalChance = chances.fractionalChances[i];
    if (fractionalChance[0].eq(fractionalChance[1])) {
      if (displayChances.length === 0) {
        displayChances.push(`guaranteed ${nhko}`);
      }
      break;
    } else if (!fractionalChance[0].isZero()) {
      const c = Math.round(chances.roundedChances[i] * 1000) / 10;
      if (c === 100) {
        displayChances.push(`almost guaranteed ${nhko}`);
      } else if (c === 0) {
        displayChances.push(`negligible chance to ${nhko}`);
      } else {
        displayChances.push(`${c}% chance to ${nhko}`);
      }
    }
  }

  if (displayChances.length > 0) {
    reportResult.push(displayChances.join(", "));
    if (effects.messages.length > 0) {
      reportResult.push("after", effects.messages.join(", "));
    }
  } else {
    reportResult.push("this might take a while...");
  }

  defender.currentHpRange = chances.remainingHealth;
  defender.currentHpRangeBerry = chances.remainingHealthBerry;

  const pokeInfo = reportPokes.join(" ");
  const damageInfo = reportDamage.join(" ");
  const resultInfo = reportResult.join(" ");

  return {
    summary: `${pokeInfo}: ${damageInfo} -- ${resultInfo}`,
    attacker,
    defender,
    move,
    field,
    minPercent,
    maxPercent,
    damages: dmg.slice(0, dmg.length - 1),
    damage: dmg[0],
    effectValues: effects.values,
    effectMessages: effects.messages,
    fractionalChances: chances.fractionalChances,
    roundedChances: chances.roundedChances
  };
}

function chanceToKo(poke, damageRanges, params) {
  const chances = [];
  const totalHp = poke.stat(Stats.HP);
  const berryHeal = poke.item.berryHeal(totalHp);
  const berryHealThreshold = poke.item.berryHealThreshold(totalHp);
  const maxTurns = params.maxTurns ?? 9;
  const rechargeTurns = params.rechargeTurns ?? 0;
  const effects = params.effects ?? [0];

  let dmg = new Multiset(params.initDmgRange ?? [0]);
  let berryDmg = new Multiset(params.initDmgRangeBerry ?? []);
  let toxicCounter = params.toxicCounter ?? 0;

  let remainingHealth, remainingHealthBerry;

  const hasFainted = damage => damage >= totalHp;
  const damageToHealth = damage => totalHp - damage;
  const applyDamage = (damage, newDamage) =>
    damage >= totalHp ? damage : damage + newDamage;

  for (let turn = 0, i = 0; turn < maxTurns; turn++, i++) {
    if (typeof damageRanges[i] === "number") {
      if (damageRanges[i] === 0) break;
      i += damageRanges[i];
    }

    dmg = dmg.permute(damageRanges[i], applyDamage).simplify();
    berryDmg = berryDmg.permute(damageRanges[i], applyDamage).simplify();

    for (let j = 0; j <= rechargeTurns; j++) {
      /*
       * because effects always has a 0 value passed in the first index,
       * berry check runs right after damage is applied, as well as after
       * every added effect. Separating damages with berries applied
       * prevents extra applications.
       */
      // berries go first to prevent double effect application
      berryDmg = berryDmg.map(
        berryDamageMap.bind({
          effects,
          toxicCounter,
          totalHp
        })
      );

      dmg = dmg.map(
        damageMap.bind({
          effects,
          toxicCounter,
          totalHp,
          berryHeal,
          berryHealThreshold,
          berryDmg
        })
      );

      toxicCounter++;
    }

    chances.push([
      dmg.count(hasFainted).add(berryDmg.count(hasFainted)),
      dmg.size.add(berryDmg.size)
    ]);

    if (turn === 0) {
      remainingHealth = dmg.map(damageToHealth);
      remainingHealthBerry = berryDmg.map(damageToHealth);
    }

    if (chances[chances.length - 1][0] === chances[chances.length - 1][1]) {
      break;
    }
  }

  return {
    fractionalChances: chances,
    roundedChances: chances.map(
      ([numerator, denominator]) =>
        numerator.multiply(1000000).divide(denominator) / 1000000
    ),
    remainingHealth,
    remainingHealthBerry
  };
}

function berryDamageMap(v) {
  for (let e = 0; e < this.effects.length && v < this.totalHp; e++) {
    if (this.effects[e] === "toxic") {
      // limit to at most enough to KO
      v += Math.trunc(((this.toxicCounter + 1) * this.totalHp) / 16);
    } else {
      // limit to at most enough to KO, at least enough to fully heal
      v = Math.max(0, v - this.effects[e]);
    }
  }
  return Math.min(this.totalHp, v);
}

function damageMap(v, w, skip) {
  let berryUsed = false;
  for (let e = 0; e < this.effects.length && v < this.totalHp; e++) {
    if (this.effects[e] === "toxic") {
      // limit to at most enough to KO
      v += Math.trunc(((this.toxicCounter + 1) * this.totalHp) / 16);
    } else {
      // limit to at most enough to KO, at least enough to fully heal
      v = Math.max(0, v - this.effects[e]);
    }
    if (
      !berryUsed &&
      this.berryHeal > 0 &&
      v < this.totalHp &&
      this.totalHp - v < this.berryHealThreshold
    ) {
      v = Math.max(0, v - this.berryHeal);
      berryUsed = true;
    }
  }
  v = Math.min(this.totalHp, v);
  // berry might not be the last effect added, so do this at the end
  if (berryUsed) {
    // separate berry modified value into berryDmg
    this.berryDmg.add(v, w);
    skip();
  }
  return v;
}
