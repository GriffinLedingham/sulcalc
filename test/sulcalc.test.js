import sulcalc from "../src/sulcalc";
import { Gens, Natures, Statuses, Terrains, maxGen } from "../src/utilities";
import { NoPokemonError, NoMoveError } from "../src/errors";

describe("sulcalc()", () => {
  test("sanity check", () => {
    for (const gen of Object.values(Gens)) {
      const { summary, damage } = sulcalc(
        { name: "Snorlax", gen },
        { name: "Snorlax", gen },
        { name: "Body Slam", gen },
        { gen }
      );
      expect(summary).toMatchSnapshot();
      expect(damage).toMatchSnapshot();
    }
  });

  test("bail out on insufficient damage", () => {
    const gen = maxGen;
    const { summary } = sulcalc(
      { name: "Munchlax", gen },
      { name: "Aggron", gen },
      { name: "Pound", gen },
      { gen }
    );
    expect(summary).toMatchSnapshot();
  });

  test("guaranteed KO", () => {
    const gen = Gens.GSC;
    const { summary } = sulcalc(
      { name: "Snorlax", gen },
      { name: "Raikou", gen },
      { name: "Self-Destruct", gen },
      { gen }
    );
    expect(summary).toMatchSnapshot();
  });

  test("almost guaranteed KO", () => {
    const gen = Gens.GSC;
    const { summary } = sulcalc(
      {
        name: "Snorlax",
        evs: [252, 200, 252, 252, 252, 252],
        gen
      },
      { name: "Snorlax", gen },
      { name: "Double-Edge", gen },
      { gen }
    );
    expect(summary).toMatchSnapshot();
  });

  test("negligible chance to KO", () => {
    const gen = Gens.GSC;
    const { summary } = sulcalc(
      {
        name: "Tauros",
        evs: [252, 176, 252, 252, 252, 252],
        gen
      },
      {
        name: "Snorlax",
        evs: [252, 252, 252, 252, 252, 252],
        gen
      },
      { name: "Body Slam", gen },
      { gen }
    );
    expect(summary).toMatchSnapshot();
  });

  test("Light Screen", () => {
    const gen = maxGen;
    const { summary } = sulcalc(
      {
        name: "Heatran",
        item: "Choice Specs",
        nature: Natures.MODEST,
        evs: [4, 0, 0, 252, 0, 252],
        gen
      },
      {
        name: "Klefki",
        nature: Natures.CALM,
        evs: [252, 0, 100, 0, 156, 0],
        lightScreen: true,
        gen
      },
      { name: "Fire Blast", gen },
      { gen }
    );
    expect(summary).toMatchSnapshot();
  });

  test("Reflect", () => {
    const gen = maxGen;
    const { summary } = sulcalc(
      {
        name: "Ferrothorn",
        nature: Natures.RELAXED,
        evs: [252, 40, 100, 0, 0, 0],
        gen
      },
      {
        name: "Starmie",
        nature: Natures.TIMID,
        evs: [4, 0, 0, 252, 0, 252],
        reflect: true,
        gen
      },
      { name: "Fire Blast", gen },
      { gen }
    );
    expect(summary).toMatchSnapshot();
  });

  test("Psyshock", () => {
    const gen = maxGen;
    const { summary } = sulcalc(
      {
        name: "Latios",
        item: "Life Orb",
        nature: Natures.TIMID,
        evs: [4, 0, 0, 252, 0, 252],
        gen
      },
      {
        name: "Chansey",
        item: "Eviolite",
        nature: Natures.BOLD,
        evs: [252, 0, 252, 0, 4, 0],
        gen
      },
      { name: "Psyshock", gen },
      { gen }
    );
    expect(summary).toMatchSnapshot();
  });

  test("Ability", () => {
    const gen = maxGen;
    const { summary } = sulcalc(
      {
        name: "Heracross",
        ability: "Guts",
        nature: Natures.ADAMANT,
        evs: [0, 252, 4, 0, 0, 252],
        boosts: [0, 2, 0, 0, 0, 0, 0, 0],
        status: Statuses.BURNED
      },
      {
        name: "Skarmory",
        ability: "Sturdy",
        nature: Natures.IMPISH,
        evs: [248, 0, 252, 0, 8, 0]
      },
      {
        name: "Close Combat",
        critical: true,
        gen
      },
      { gen }
    );
    expect(summary).toMatchSnapshot();
  });

  test("Explosion", () => {
    const gen = maxGen;
    const { roundedChances } = sulcalc(
      { name: "Snorlax", gen },
      { name: "Snorlax", gen },
      { name: "Explosion", gen },
      { gen }
    );
    expect(roundedChances.length).toEqual(1);
    const { summary } = sulcalc(
      { name: "Snorlax", gen },
      { name: "Skarmory", gen },
      { name: "Explosion", gen },
      { gen }
    );
    expect(summary).toMatchSnapshot();
  });

  test("throws NoPokemonError when either Pokemon is invalid", () => {
    expect(() => {
      sulcalc({ id: "0:0" }, { id: "0:0" }, { id: 1 }, {});
    }).toThrow(NoPokemonError);

    expect(() => {
      sulcalc({ id: "233" }, { id: "233" }, { id: 1 }, {});
    }).toThrow(NoPokemonError);
  });

  test("throws NoMoveError when either Move is invalid", () => {
    expect(() => {
      sulcalc({ name: "Snorlax" }, { name: "Snorlax" }, { id: 0 }, {});
    }).toThrow(NoMoveError);

    expect(() => {
      sulcalc({ name: "Snorlax" }, { name: "Snorlax" }, { id: -1 }, {});
    }).toThrow(NoMoveError);
  });

  test("Variable BP fix", () => {
    const gen = Gens.GSC;
    const { summary } = sulcalc(
      {
        name: "Gyarados",
        item: "Polkadot Bow",
        currentHp: 15,
        gen
      },
      { name: "Articuno", gen },
      { name: "Flail", gen },
      { gen }
    );
    expect(summary).toMatchSnapshot();
  });

  test("Present GSC", () => {
    const gen = Gens.GSC;
    for (const defenderName of ["Raikou", "Vaporeon", "Zapdos"]) {
      for (let present = -1; present <= 3; present++) {
        expect(
          sulcalc(
            { name: "Blissey", gen },
            { name: defenderName, gen },
            { name: "Present", present, gen },
            { gen }
          ).summary
        ).toMatchSnapshot();
      }
    }
  });

  test("Terrains boost types", () => {
    for (let gen = Gens.ORAS; gen <= maxGen; gen++) {
      expect(
        sulcalc(
          {
            name: "Tapu Lele",
            nature: Natures.TIMID,
            evs: [0, 0, 0, 252, 4, 252],
            gen
          },
          {
            name: "Landorus-Therian",
            nature: Natures.ADAMANT,
            evs: [0, 252, 4, 0, 0, 252],
            gen
          },
          { name: "Psychic", gen },
          { terrain: Terrains.PSYCHIC_TERRAIN, gen }
        ).summary
      ).toMatchSnapshot();

      expect(
        sulcalc(
          {
            name: "Tapu Koko",
            nature: Natures.NAIVE,
            evs: [0, 252, 0, 4, 0, 252],
            gen
          },
          {
            name: "Diancie",
            nature: Natures.NAIVE,
            evs: [0, 4, 0, 252, 0, 252],
            gen
          },
          { name: "Wild Charge", gen },
          { terrain: Terrains.ELECTRIC_TERRAIN, gen }
        ).summary
      ).toMatchSnapshot();

      expect(
        sulcalc(
          {
            name: "Tapu Bulu",
            item: "Choice Band",
            nature: Natures.ADAMANT,
            evs: [0, 252, 4, 0, 0, 252],
            boosts: [0, -1, 0, 0, 0, 0, 0, 0],
            gen
          },
          {
            name: "Landorus-Therian",
            nature: Natures.IMPISH,
            evs: [252, 0, 216, 0, 24, 16],
            gen
          },
          { name: "Wood Hammer", gen },
          { terrain: Terrains.GRASSY_TERRAIN, gen }
        ).summary
      ).toMatchSnapshot();
    }
  });

  test("Greninja-Ash Water Shuriken", () => {
    const gen = maxGen;
    expect(
      sulcalc(
        {
          name: "Greninja-Ash",
          item: "Choice Specs",
          nature: Natures.TIMID,
          evs: [0, 0, 4, 252, 0, 252],
          gen
        },
        {
          name: "Tapu Koko",
          nature: Natures.NAIVE,
          evs: [0, 252, 0, 4, 0, 252],
          gen
        },
        { name: "Water Shuriken", gen },
        { gen }
      ).summary
    ).toMatchSnapshot();
  });

  test("Skill Link", () => {
    const gen = maxGen;
    expect(
      sulcalc(
        {
          name: "Cloyster",
          nature: Natures.NAUGHTY,
          ability: "Skill Link",
          evs: [0, 252, 4, 0, 0, 252],
          gen
        },
        {
          name: "Rotom-Wash",
          nature: Natures.BOLD,
          evs: [248, 0, 200, 0, 0, 60],
          gen
        },
        { name: "Icicle Spear", gen },
        { gen }
      ).summary
    ).toMatchSnapshot();
  });

  test("Thousand Arrows", () => {
    const gen = maxGen;
    expect(
      sulcalc(
        {
          name: "Zygarde",
          item: "Choice Band",
          nature: Natures.ADAMANT,
          evs: [4, 252, 0, 0, 0, 252],
          gen
        },
        {
          name: "Rotom-Wash",
          nature: Natures.BOLD,
          ability: "Levitate",
          evs: [248, 0, 200, 0, 0, 60],
          gen
        },
        { name: "Thousand Arrows", gen },
        { gen }
      ).summary
    ).toMatchSnapshot();
  });

  test("Levitate immunity to Ground moves", () => {
    const gen = maxGen;
    expect(
      sulcalc(
        {
          name: "Zygarde",
          item: "Choice Band",
          nature: Natures.ADAMANT,
          evs: [4, 252, 0, 0, 0, 252],
          gen
        },
        {
          name: "Rotom-Wash",
          nature: Natures.BOLD,
          ability: "Levitate",
          evs: [248, 0, 200, 0, 0, 60],
          gen
        },
        { name: "Earthquake", gen },
        { gen }
      ).summary
    ).toMatchSnapshot();
  });

  test("Grassy Terrain recovery", () => {
    const gen = maxGen;
    expect(
      sulcalc(
        {
          name: "Tapu Bulu",
          nature: Natures.JOLLY,
          evs: [0, 252, 4, 0, 0, 252],
          gen
        },
        {
          name: "Chansey",
          item: "Eviolite",
          nature: Natures.BOLD,
          evs: [248, 0, 252, 0, 8, 0],
          gen
        },
        { name: "Horn Leech", gen },
        { terrain: Terrains.GRASSY_TERRAIN, gen }
      ).summary
    ).toMatchSnapshot();
  });

  test("Defender health text shows for not completely healthy Pokemon", () => {
    const gen = Gens.RBY;
    expect(
      sulcalc(
        { name: "Tauros", gen },
        { name: "Snorlax", currentHp: 40, gen },
        { name: "Hyper Beam", gen },
        { gen }
      ).summary
    ).toContain("at 7%");
  });
});
