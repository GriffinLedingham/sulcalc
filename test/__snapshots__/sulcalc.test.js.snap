// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sulcalc() Ability 1`] = `"+2 252+ Atk Guts Heracross (burned) critical hit Close Combat vs. 248 HP / 252+ Def Sturdy Skarmory: 332 (99.7%) -- guaranteed 2HKO"`;

exports[`sulcalc() Explosion 1`] = `"0 Atk Snorlax Explosion vs. 0 HP / 0 Def Skarmory: 109 - 129 (40.2 - 47.6%) -- this might take a while..."`;

exports[`sulcalc() Light Screen 1`] = `"252+ SpA Choice Specs Heatran Fire Blast vs. 252 HP / 156+ SpD Klefki behind Light Screen: 256 - 303 (80.5 - 95.3%) -- guaranteed 2HKO"`;

exports[`sulcalc() Psyshock 1`] = `"252 SpA Life Orb Latios Psyshock vs. 252 HP / 252+ Def Eviolite Chansey: 226 - 266 (32.1 - 37.8%) -- 93.8% chance to 3HKO"`;

exports[`sulcalc() Reflect 1`] = `"0 SpA Ferrothorn Fire Blast vs. 4 HP / 0 SpD Starmie: 28 - 33 (10.7 - 12.6%) -- negligible chance to 8HKO, 98.7% chance to 9HKO"`;

exports[`sulcalc() Variable BP fix 1`] = `"Polkadot Bow Gyarados Flail vs. Articuno: 218 (56.9%) -- guaranteed 2HKO"`;

exports[`sulcalc() almost guaranteed KO 1`] = `"200 Atk Snorlax Double-Edge vs. Snorlax: 173 - 204 (33.1 - 39%) -- almost guaranteed 3HKO"`;

exports[`sulcalc() bail out on insufficient damage 1`] = `"0 Atk Munchlax Pound vs. 0 HP / 0 Def Aggron: 6 - 7 (2.1 - 2.5%) -- this might take a while..."`;

exports[`sulcalc() guaranteed KO 1`] = `"Snorlax Self-Destruct vs. Raikou: 548 - 645 (143.1 - 168.4%) -- guaranteed OHKO"`;

exports[`sulcalc() negligible chance to KO 1`] = `"176 Atk Tauros Body Slam vs. Snorlax: 112 - 132 (21.4 - 25.2%) -- negligible chance to 4HKO"`;

exports[`sulcalc() sanity check 1`] = `"Snorlax Body Slam vs. Snorlax: 127 - 150 (24.3 - 28.7%) -- 99.2% chance to 4HKO"`;

exports[`sulcalc() sanity check 2`] = `
Multiset {
  "_data": Map {
    127 => "1",
    128 => "2",
    129 => "1",
    130 => "2",
    131 => "2",
    132 => "2",
    133 => "1",
    134 => "2",
    135 => "2",
    136 => "1",
    137 => "2",
    138 => "2",
    139 => "1",
    140 => "2",
    141 => "2",
    142 => "2",
    143 => "1",
    144 => "2",
    145 => "2",
    146 => "1",
    147 => "2",
    148 => "2",
    149 => "1",
    150 => "1",
  },
}
`;

exports[`sulcalc() sanity check 3`] = `"Snorlax Body Slam vs. Snorlax: 127 - 150 (24.3 - 28.7%) -- 99.2% chance to 4HKO"`;

exports[`sulcalc() sanity check 4`] = `
Multiset {
  "_data": Map {
    127 => "1",
    128 => "2",
    129 => "1",
    130 => "2",
    131 => "2",
    132 => "2",
    133 => "1",
    134 => "2",
    135 => "2",
    136 => "1",
    137 => "2",
    138 => "2",
    139 => "1",
    140 => "2",
    141 => "2",
    142 => "2",
    143 => "1",
    144 => "2",
    145 => "2",
    146 => "1",
    147 => "2",
    148 => "2",
    149 => "1",
    150 => "1",
  },
}
`;

exports[`sulcalc() sanity check 5`] = `"0 Atk Snorlax Body Slam vs. 0 HP / 0 Def Snorlax: 142 - 168 (30.8 - 36.4%) -- 61.1% chance to 3HKO"`;

exports[`sulcalc() sanity check 6`] = `
Multiset {
  "_data": Map {
    142 => "1",
    144 => "1",
    146 => "1",
    147 => "1",
    149 => "1",
    151 => "1",
    152 => "1",
    154 => "1",
    156 => "1",
    157 => "1",
    159 => "1",
    161 => "1",
    162 => "1",
    164 => "1",
    166 => "1",
    168 => "1",
  },
}
`;

exports[`sulcalc() sanity check 7`] = `"0 Atk Snorlax Body Slam vs. 0 HP / 0 Def Snorlax: 142 - 168 (30.8 - 36.4%) -- 57.8% chance to 3HKO"`;

exports[`sulcalc() sanity check 8`] = `
Multiset {
  "_data": Map {
    142 => "1",
    144 => "1",
    145 => "1",
    147 => "1",
    148 => "1",
    150 => "1",
    151 => "1",
    154 => "1",
    156 => "1",
    157 => "1",
    159 => "1",
    160 => "1",
    162 => "1",
    163 => "1",
    165 => "1",
    168 => "1",
  },
}
`;

exports[`sulcalc() sanity check 9`] = `"0 Atk Snorlax Body Slam vs. 0 HP / 0 Def Snorlax: 142 - 168 (30.8 - 36.4%) -- 57.8% chance to 3HKO"`;

exports[`sulcalc() sanity check 10`] = `
Multiset {
  "_data": Map {
    142 => "1",
    144 => "1",
    145 => "1",
    147 => "1",
    148 => "1",
    150 => "1",
    151 => "1",
    154 => "1",
    156 => "1",
    157 => "1",
    159 => "1",
    160 => "1",
    162 => "1",
    163 => "1",
    165 => "1",
    168 => "1",
  },
}
`;

exports[`sulcalc() sanity check 11`] = `"0 Atk Snorlax Body Slam vs. 0 HP / 0 Def Snorlax: 142 - 168 (30.8 - 36.4%) -- 57.8% chance to 3HKO"`;

exports[`sulcalc() sanity check 12`] = `
Multiset {
  "_data": Map {
    142 => "1",
    144 => "1",
    145 => "1",
    147 => "1",
    148 => "1",
    150 => "1",
    151 => "1",
    154 => "1",
    156 => "1",
    157 => "1",
    159 => "1",
    160 => "1",
    162 => "1",
    163 => "1",
    165 => "1",
    168 => "1",
  },
}
`;

exports[`sulcalc() sanity check 13`] = `"0 Atk Snorlax Body Slam vs. 0 HP / 0 Def Snorlax: 142 - 168 (30.8 - 36.4%) -- 57.8% chance to 3HKO"`;

exports[`sulcalc() sanity check 14`] = `
Multiset {
  "_data": Map {
    142 => "1",
    144 => "1",
    145 => "1",
    147 => "1",
    148 => "1",
    150 => "1",
    151 => "1",
    154 => "1",
    156 => "1",
    157 => "1",
    159 => "1",
    160 => "1",
    162 => "1",
    163 => "1",
    165 => "1",
    168 => "1",
  },
}
`;
